bitcoin/src/blockdata/script/mod.rs

```diff
@@ -680,6 +736,148 @@ pub(super) fn bytes_to_asm_fmt(script: &[u8], f: &mut dyn fmt::Write) -> fmt::Re
     Ok(())
 }
 
+/// The different kinds of [ParseAsmError] that can occur.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ParseAsmErrorKind {
+    /// ASM ended unexpectedly.
+    UnexpectedEOF,
+    /// We were not able to interpret the instruction.
+    UnknownInstruction,
+    /// Invalid hexadecimal bytes.
+    InvalidHex,
+    /// Byte push exceeding the maximum size.
+    PushExceedsMaxSize,
+    /// ASM contains a byte push with non-minimal size prefix.
+    ///
+    /// This is not necessarily invalid, but we can't construct such pushes.
+    NonMinimalBytePush,
+}
+
+/// Error from parsing Script ASM.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ParseAsmError {
+    /// The position of the instruction that caused the error.
+    ///
+    /// The value is (line, word) with word incremented after
+    /// every chunk of whitespace.
+    pub position: (usize, usize),
+
+    /// The kind of error that occurred.
+    pub kind: ParseAsmErrorKind,
+}
+
+/// Try to parse raw hex bytes and push them into the buffer.
+fn try_parse_raw_hex(hex: &str, buf: &mut Vec<u8>) -> bool {
+    buf.clear();
+    let iter = match hex::HexToBytesIter::new(hex) {
+        Ok(i) => i,
+        Err(_) => return false,
+    };
+    for item in iter {
+        let item = match item {
+            Ok(i) => i,
+            Err(_) => return false,
+        };
+        buf.push(item);
+    }
+    true
+}
+
+/// Create an iterator over instruction words and their position in the file.
+fn iter_words(asm: &str) -> impl Iterator<Item = ((usize, usize), &str)> {
+    asm.lines().enumerate().flat_map(|(line_idx, line)| {
+        let content = line.split('#').next().unwrap().split("//").next().unwrap();
+        content.split_whitespace().enumerate().map(move |(word_idx, word)| {
+            ((line_idx, word_idx), word)
+        })
+    })
+}
+
+/// Parse a Script in ASM format.
+pub(super) fn parse_asm(asm: &str) -> Result<ScriptBuf, ParseAsmError> {
+    fn err(position: (usize, usize), kind: ParseAsmErrorKind) -> ParseAsmError {
+        ParseAsmError { position, kind }
+    }
+
+    let mut buf = Vec::with_capacity(65);
+    let mut builder = Builder::new();
+    let mut words = iter_words(asm);
+    while let Some((pos, mut word)) = words.next() {
+        // We have this special case in our formatter.
+        if word == "OP_0" {
+            builder = builder.push_opcode(OP_PUSHBYTES_0);
+            continue;
+        }
+
+        if let Ok(op) = Opcode::from_str(word) {
+            // check for push opcodes
+            if op.code <= OP_PUSHDATA4.code {
+                let (next, push) = words.next().ok_or(err(pos, ParseAsmErrorKind::UnexpectedEOF))?;
+                if !try_parse_raw_hex(push, &mut buf) {
+                    return Err(err(next, ParseAsmErrorKind::InvalidHex));
+                }
```

Comment:
We can get rid of `try_parse_raw_hex` and the mutable buffer and just use:
```rust
let v = Vec::from_hex(push).map_err(|_| err(next, ParseAsmErrorKind::InvalidHex))?;
```

Reply:
Does using `Vec::from_hex` instead of a mutable buffer introduce an allocation where previously there wasn't one?